cclient/api/extended/get_inputs.c

```diff
@@ -4,63 +4,130 @@
-        tmp_input.balance = get_balances_res_balances_at(balances_res, counter);
-        tmp_input.key_index = counter;
+  out_inputs->total_balance = 0;
+
+  flex_trit_t* addr = NULL;
+  bool is_unused = false;
+  uint64_t addr_index, addr_index_end;
+  if (addr_opt.total == 0) {
+    addr_index = 0;
+    addr_index_end = (unsigned)-1;
+  } else {
+    addr_index = addr_opt.start;
+    addr_index_end = addr_opt.total;
+  }
+
+  for (; addr_index < addr_index_end; addr_index++) {
+    balances_req = get_balances_req_new();
+    balances_res = get_balances_res_new();
+    log_debug(client_extended_logger_id, "[%s:%d]\n", __func__, __LINE__);
+    if (!balances_req || !balances_res) {
+      ret_code = RC_OOM;
+      log_error(client_extended_logger_id, "%s create get balances request or response object failed: %s\n", __func__,
+                error_2_string(ret_code));
+      goto done;
+    }
+
+    addr = iota_sign_address_gen_flex_trits(seed, addr_index, addr_opt.security);
+    if (addr) {
+      ret_code = hash243_queue_push(&balances_req->addresses, addr);
+      if (ret_code) {
+        log_error(client_extended_logger_id, "%s:%d hash queue push failed: %s\n", __func__, __LINE__,
+                  error_2_string(ret_code));
+        goto done;
+      }
+    } else {
+      // gen address failed.
+      ret_code = RC_OOM;
+      log_error(client_extended_logger_id, "%s address generation failed: %s\n", __func__, error_2_string(ret_code));
+      goto done;
+    }
+    
+    // For request without `addr_opt.total`, end this loop if the current addr is an unused address
+    if (addr_opt.total == 0) {
+      ret_code = is_unused_address(serv, addr, &is_unused);
+      if (ret_code == RC_OK && is_unused) {
+        break;
+      }
+    }
+    
+    balances_req->threshold = 100;  // currently `100` should be used
+
+    ret_code = iota_client_get_balances(serv, balances_req, balances_res);
+    if (ret_code == RC_OK && addr) {
+      // expect balance value is in order.
+      if (out_inputs->total_balance < threshold) {
+        input_t tmp_input = {};
+        tmp_input.balance = get_balances_res_balances_at(balances_res, 0);
+        tmp_input.key_index = addr_index;
         tmp_input.security = addr_opt.security;
-        memcpy(tmp_input.address, q_iter->hash, FLEX_TRIT_SIZE_243);
+        memcpy(tmp_input.address, addr, FLEX_TRIT_SIZE_243);
         if ((ret_code = inputs_append(out_inputs, &tmp_input)) != RC_OK) {
           log_error(client_extended_logger_id, "%s get new address failed: %s\n", __func__, error_2_string(ret_code));
           goto done;
         }
+      } else {
+        break;
       }
-      counter++;
-    }
-    // check if balance is sufficient
-    if (out_inputs->total_balance < (int64_t)threshold) {
-      ret_code = RC_CCLIENT_INSUFFICIENT_BALANCE;
-      log_warning(client_extended_logger_id, "insufficient balance\n");
     }
+
+    get_balances_req_free(&balances_req);
```

Comment:
you can reuse the request and response instead of allocating every time.

Reply:
`hash243_queue_to_json_array` would convert all the elements in `hash243_queue_t` to json array,  so if I didn't free the `hash243_queue_t` variable, then `hash243_queue_to_json_array` would generate a wrong json array.
In the other hand, if reuse them without freeing the `get_balances_req_t` and `get_balances_res_t` object, then there will be memory leak problems, since the elements such as `hash243_queue_t` should be freed.

Reply:
you can clean the queue instead freeing the entire struct, isn't it?

Reply:
sure